docs/design/api-versions.md

```diff
@@ -32,8 +40,169 @@ For each API spec, generated `AssignPropertiesTo*()` method  will read the value
 
 For each Storage spec, the generated `AssignPropertiesTo*()` and `AssignPropertiesFrom*()` methods will copy the `OriginalVersion` property between versions, preserving the original value.
 
-### API Recovery
-
-Into each storage resource variant, we'll inject a function `OriginalGVK(): GroupVersionKind` which will use `OriginalVersion` to create the GVK required.
+## New conversion interfaces
 
-SequenceDiagram TBC.
+We don't want the overhead of doing a complete conversion of each resource if we're only interested in conversion of the associated `Spec` or `Status`. For example, when submitting a request to ARM we need only do a `Spec` conversion; similiarly, when processing a response from ARM, we need only do a `Status` conversion.
+
+We already have the required infrastructure in the form of the `AssignProperties*()` methods, making it simple to implement these new interfaces.
+
+For conversion of `Spec` types:
+
+``` go
+type ConvertibleSpec interface {
+    // ConvertFromSpec populates the current spec from the passed one
+    ConvertFromSpec(source ConvertibleSpec) error
+    
+    // ConvertToSpec populates the passed spec from the current one
+    ConvertToSpec(destination ConvertibleSpec) error
+}
+```
+
+As we have done with implementations of the traditional [`Convertible`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/conversion#Convertible) interface, these will be implemented on Specs that are further from our hub, and will chain towards that hub.
+
+For conversion of `Status` types, a similar approach:
+
+``` go
+type ConvertibleStatus interface {
+    // ConvertFromStatus populates the current status from the passed one
+    ConvertFromStatus(source ConvertibleStatus) error
+    
+    // ConvertToStatus populates the passed status from the current one
+    ConvertToStatus(destination ConvertibleStatus) error
+}
+```
+
+## Reconciler integration
+
+When the reconciler is working with ARM, we'll use the `OriginalVersion` property on the storage spec to pivot from the current version to the original version.
+
+To keep the reconciler a simple as possible, we'll handle this with generated code hidden behind a simple interface:
+
+``` go
+type GeneratedResource interface {
+    // GetSpec returns the spec of the resource ready for 
+    // serialization to ARM
+    GetSpec(scheme runtime.Scheme) (genruntime.ARMTransformer, error)
```

Comment:
In keeping with the above, if this were on a `KubernetesResource` interface maybe this is `GetSpecARMConverter` or something to differentiate from another possible method `GetSpec()` which just returned the spec as a `runtime.Object` or something?

Reply:
Same with the below methods, consider sprinkling some `ARM` specific stuff in the name to contrast with another possible method which creates the corresponding Kubernetes type rather than the ARM type.

Reply:
These methods don't return ARM types; they return implementations that can create ARM types.

Reply:
I suggested `GetARMSpecConverter` not `GetARMSpec` - I agree that `GetARMSpec` would be misleading, but I also think that `GetSpec` is misleading because this isn't really returning a spec, it's returning a thing that can be used to make an ARM type from a spec, and `GetSpec` seems better reserved for a method that actually returns the spec itself (and has no required parameters, unlike the required `scheme` here)

Reply:
I'm moving the conversion into the reconciler - which both removes the need for a `scheme` parameter, and avoids confusion here because `GetSpec()` just returns the spec on the resource, no conversion required.